test/jaxp/javax/xml/jaxp/unittest/common/jdkcatalog/JDKCatalogTest.java

```diff
@@ -0,0 +1,275 @@
+                targetNS, rootElement);
+        if (expectedThrow == null) {
+            assertDoesNotThrow(() -> validateWithResolveStrict(xmlString, xmlSrcId, customCatalog),
+                    "JDKCatalog shall resolve " + xsdLocation + " but exception is thrown.");
+        } else {
+            Assert.assertThrows(expectedThrow,
+                () -> validateWithResolveStrict(xmlString, xmlSrcId, customCatalog));
+        }
+    }
+
+    /**
+     * Validate the specified XML document with jdk.xml.jdkCatalog.resolve set to strict.
+     * @param xml  the XML document to be validated
+     * @param xmlSrcPathId the URI to the XML source (template in this case)
+     * @param customCatalog the custom catalog used to resolve local XSDs
+     * @throws Exception if validation fails
+     */
+    public void validateWithResolveStrict(String xml, String xmlSrcPathId, String customCatalog)
+            throws Exception {
+        SAXSource ss = new SAXSource(new InputSource(new StringReader(xml)));
+        ss.setSystemId(xmlSrcPathId);
+        SchemaFactory schemaFactory = SchemaFactory.newInstance(XMLConstants.W3C_XML_SCHEMA_NS_URI);
+        schemaFactory.setProperty(JDKCATALOG_RESOLVE, "strict");
+        if (customCatalog != null) {
+            schemaFactory.setProperty(CatalogFeatures.Feature.FILES.getPropertyName(), customCatalog);
+            schemaFactory.setProperty(CatalogFeatures.Feature.RESOLVE.getPropertyName(), "continue");
+        }
+        Validator validator = schemaFactory.newSchema().newValidator();
+        validator.validate(ss);
+    }
+
+    /**
+     * Parses the XML with jdk.xml.jdkCatalog.resolve set to strict.
+     * @param xml the XML document to be parsed
+     * @throws Exception if external access is denied
+     */
+    public void parseWithResolveStrict(String xml)
+            throws Exception {
+        SAXParserFactory saxParserFactory = SAXParserFactory.newInstance();
+        saxParserFactory.setNamespaceAware(true);
+        XMLReader xmlReader = saxParserFactory.newSAXParser().getXMLReader();
+        xmlReader.setProperty(JDKCATALOG_RESOLVE, "strict");
+        xmlReader.setContentHandler(new DefaultHandler());
+        xmlReader.parse(new InputSource(new StringReader(xml)));
+    }
+
+    /**
+     * Generates an XML with the specified PUBLIC and SYSTEM identifiers.
+     * @param publicId the public identifier
+     * @param systemId the system identifier
+     * @return an XML
+     * @throws Exception if error happens
+     */
+    private String generateXMLWithDTDRef(String publicId, String systemId)
+            throws Exception {
+        Path path = Paths.get(SRC_DIR + "/dtdtest.xml");
+        String xmlString = Files.lines(path).map(line -> {
+            if (line.contains(PUBLIC_ID)) {
+                line = line.replace(PUBLIC_ID, publicId);
+            }
+            if (line.contains(SYSTEM_ID)) {
+                line = line.replace(SYSTEM_ID, systemId);
+            }
```

Comment:
The `if ... contains ...` isn't necessary; String.replace is a no-op if the string is not present.